src/sinks/elasticsearch.rs

```diff
@@ -306,47 +306,26 @@ impl RetryLogic for ElasticSearchRetryLogic {
             )),
             _ if status.is_client_error() => {
                 let body = String::from_utf8_lossy(response.body());
-                warn!(
-                    message = "Client error",
-                    body = %body,
-                    rate_limit_secs = 30
-                );
-                RetryAction::DontRetry("client error".into())
+                RetryAction::DontRetry(format!("client-side error, {}: {}", status, body))
             }
             _ if status.is_success() => {
                 let body = String::from_utf8_lossy(response.body());
-                match body.find("\"errors\":true") {
-                    Some(_) => match serde_json::from_str::<ESResultResponse>(&body) {
-                        Err(json_error) => {
-                            warn!(
-                                message = "Elasticsearch unparsable error response",
-                                %json_error,
-                                rate_limit_secs = 30
-                            );
-                            RetryAction::DontRetry(
-                                "Some messages failed, and invalid response from Elasticsearch"
-                                    .into(),
-                            )
-                        }
-                        Ok(esrr) => {
-                            match esrr.items.into_iter().find_map(|item| item.index.error) {
-                                Some(error) => warn!(
-                                    message = "ElasticSearch error response",
-                                    err_type = %error.err_type,
-                                    reason = %error.reason,
-                                    rate_limit_secs = 30
-                                ),
-                                _ => warn!(
-                                    message = "Unusual ElasticSearch error response",
-                                    %body,
-                                    rate_limit_secs = 30
-                                ),
-                            };
-                            RetryAction::DontRetry("some messages failed".into())
+
+                if !body.contains("\"errors\":true") {
+                    return RetryAction::Successful;
+                }
+
+                let reason = match serde_json::from_str::<ESResultResponse>(&body) {
+                    Err(json_error) => format!("could not parse response, error: {}", json_error),
+                    Ok(resp) => match resp.items.into_iter().find_map(|item| item.index.error) {
+                        Some(error) => {
+                            format!("error type: {}, reason: {}", error.err_type, error.reason)
                         }
+                        None => format!("error response: {}", body),
                     },
-                    None => RetryAction::Successful,
-                }
+                };
+
+                RetryAction::DontRetry(reason)
```

Comment:
This can be done without a `return`:
```rust
if body.contains("\"errors":true") {\
    …
} else {
    RetryAction::Successful
}
```

Reply:
I used early return to avoid nesting too many matches.

Reply:
I think I will persist with my objection to this one. Since this function is somewhat long both with or without the early return, consider moving out some of the nested matches into a function.

Reply:
added `get_err_reason`

Reply:
This looks excellent.